src/goto-instrument/contracts/dynamic-frames/dfcc.cpp

```diff
@@ -0,0 +1,497 @@
+  const bool apply_loop_contracts,
+  const std::set<std::string> &to_exclude_from_nondet_static)
+{
+  // check there is at most one function to check
+  PRECONDITION_WITH_DIAGNOSTICS(
+    to_check.size() <= 1,
+    "Only a single (function, contract) pair can be checked at a time, "
+    "aborting");
+
+  // check that harness function exists
+  PRECONDITION_WITH_DIAGNOSTICS(
+    utils.function_symbol_with_body_exists(harness_id),
+    "Harness function '" + id2string(harness_id) +
+      "' either not found or has no body");
+
+  for(const auto &pair : to_check)
+  {
+    PRECONDITION_WITH_DIAGNOSTICS(
+      utils.function_symbol_with_body_exists(pair.first),
+      "Function to check '" + id2string(pair.first) +
+        "' either not found or has no body");
+
+    PRECONDITION_WITH_DIAGNOSTICS(
+      dsl_contract_handler.pure_contract_symbol_exists(pair.second),
+      "Contract to check '" + id2string(pair.second) + "' not found");
+
+    PRECONDITION_WITH_DIAGNOSTICS(
+      pair.first != harness_id,
+      "Function '" + id2string(pair.first) +
+        "' cannot be both be checked against a contract and be the harness");
+
+    PRECONDITION_WITH_DIAGNOSTICS(
+      pair.second != harness_id,
+      "Function '" + id2string(pair.first) +
+        "' cannot be both the contract to check and be the harness");
+
+    PRECONDITION_WITH_DIAGNOSTICS(
+      to_replace.find(pair.first) == to_replace.end(),
+      "Function '" + id2string(pair.first) +
+        "' cannot be both checked against contract and replaced by a contract");
+
+    PRECONDITION_WITH_DIAGNOSTICS(
+      !instrument.do_not_instrument(pair.first),
+      "CPROVER function or builtin '" + id2string(pair.first) +
+        "' cannot be checked against a contract");
+  }
+
+  for(const auto &pair : to_replace)
+  {
+    // for functions to replace with contracts we don't require the replaced
+    // function to have a body because only the contract is actually used
+    PRECONDITION_WITH_DIAGNOSTICS(
+      utils.function_symbol_exists(pair.first),
+      "Function to replace '" + id2string(pair.first) + "' not found");
+
+    PRECONDITION_WITH_DIAGNOSTICS(
+      dsl_contract_handler.pure_contract_symbol_exists(pair.second),
+      "Contract to replace '" + id2string(pair.second) + "' not found");
+
+    PRECONDITION_WITH_DIAGNOSTICS(
+      pair.first != harness_id,
+      "Function '" + id2string(pair.first) +
+        "' cannot both be replaced with a contract and be the harness");
+
+    PRECONDITION_WITH_DIAGNOSTICS(
+      pair.second != harness_id,
+      "Function '" + id2string(pair.first) +
+        "' cannot both be the contract to use for replacement and be the "
+        "harness");
+  }
+}
+
+void dfcct::partition_function_symbols(
+  std::set<irep_idt> &contract_symbols,
+  std::set<irep_idt> &other_symbols)
+{
+  auto &symbol_table = goto_model.symbol_table;
+
+  // collect contract and other sybols
+  for(auto &entry : symbol_table)
```

Comment:
Nit pick: The introduction of a local variable `symbol_table` seems unnecessary.

Reply:
fixed